### ISSUE TITLE
HUGE memory leak problem

### ISSUE THREAD

**Issue by jakeleventhal** [googleapis/nodejs-firestore#768]
#### Environment details

  - OS: Mac OS 10.15 Beta (19A573a)
  - Node.js version: v12.10.0
  - npm version: 6.10.3
  - `@google-cloud/firestore` version: 2.3.0

I'm still experiencing a memory leak potentially related to #661. The following is from my yarn.lock:
```
"@grpc/grpc-js@^0.5.2":
  version "0.5.2"
  resolved "https://registry.yarnpkg.com/@grpc/grpc-js/-/grpc-js-0.5.2.tgz#8e39876033749e1611fe9fa181d9535077805abe"
  integrity sha512-NE1tP/1AF6BqhLdILElnF7aOBfoky+4ZOdZU/0NmKo2d+9F9QD8zGoElpBk/5BfyQZ3u1Zs+wFbDOFpVUzDx1w==
  dependencies:
    semver "^6.0.0"

"@grpc/proto-loader@^0.5.0", "@grpc/proto-loader@^0.5.1":
  version "0.5.1"
  resolved "https://registry.yarnpkg.com/@grpc/proto-loader/-/proto-loader-0.5.1.tgz#48492b53cdda353110b51a4b02f465974729c76f"
  integrity sha512-3y0FhacYAwWvyXshH18eDkUI40wT/uGio7MAegzY8lO5+wVsc19+1A7T0pPptae4kl7bdITL+0cHpnAPmryBjQ==
  dependencies:
    lodash.camelcase "^4.3.0"
    protobufjs "^6.8.6"
```
_Quick note: because of the grpc problems that keep appearing here, we should probably put grpc version as part of the issue template_

So it is clear that I'm actually using the "correct" grpc. My application makes many calls to firestore and I'm experiencing massive memory leaks.
<img width="1673" alt="Screen Shot 2019-09-28 at 6 53 37 PM" src="https://user-images.githubusercontent.com/15133828/65825228-779fab00-e229-11e9-95c5-1c6b2c922ab3.png">

I'm not actually seeing the error message described previously (`MaxListenersExceededWarning`), however, the memory usage on my application slowly and steady is increasing. I've taken several heap snapshots from local tests over the course of several hours and when comparing snapshots from a few hours apart, I notice that the top culprits for memory allocations are all related to grpc.

Here you can clearly see that the top 4 culprits for the memory delta between snapshots 12 and 4 are vastly more than everything else
<img width="857" alt="Screen Shot 2019-09-28 at 7 08 24 PM" src="https://user-images.githubusercontent.com/15133828/65824794-59cf4780-e223-11e9-88d4-3b0838aac54f.png">

Here are the contents of each from largest to smallest:

1. (string): This by far is the fastest growing constructor. As you can see, there are 920k new allocations in the comparison between the two snapshots. Practically every single string here is `""(function anonymous(
) {
return function BeginTransactionRequest(p){
  if(p)for(var ks=Object.keys(p),i=0;i<ks.length;++i)if(p[ks[i]]!=null)
  this[ks[i]]=p[ks[i]]
}
})"`. And this all comes from `message.js:13`. This is a a file from the package `protobufjs` which is a dependency of grpc.
<img width="1450" alt="Screen Shot 2019-09-28 at 7 47 17 PM" src="https://user-images.githubusercontent.com/15133828/65825180-c8fb6a80-e228-11e9-952b-4400fb90c1f9.png">

2. (system): Essentially everything here seems to be coming from from the same `message.js:13`.
<img width="1587" alt="Screen Shot 2019-09-28 at 7 29 51 PM" src="https://user-images.githubusercontent.com/15133828/65825014-7faa1b80-e226-11e9-80b7-f05ec9ac5071.png">
I also have tens of thousands of objects that look like this (from grpc)
<img width="1154" alt="Screen Shot 2019-09-28 at 8 10 43 PM" src="https://user-images.githubusercontent.com/15133828/65825477-36f56100-e22c-11e9-9a3a-5b432c315114.png">

3. (array): Practically every single array allocation here is a alloc size 32 and is linked to `type.js:31` and `namespace.js:95`, both of which are also part of protobufjs.
<img width="1431" alt="Screen Shot 2019-09-28 at 8 18 42 PM" src="https://user-images.githubusercontent.com/15133828/65825569-2b566a00-e22d-11e9-83db-8246a84c5346.png">
"Field" is also related to the same line numbers and is actually directly linked to grpc
<img width="1416" alt="Screen Shot 2019-09-28 at 8 20 38 PM" src="https://user-images.githubusercontent.com/15133828/65825584-76707d00-e22d-11e9-9e60-7370ee9b1574.png">

4. Array: There are 235k of these and they all look the same - Array with an alloc size of 32 with some connection to grpc
<img width="1600" alt="Screen Shot 2019-09-28 at 7 12 35 PM" src="https://user-images.githubusercontent.com/15133828/65824827-ff82b680-e223-11e9-8283-919e27cf25dd.png">

**Comment by jakeleventhal** [comment]
Update: I tried updating to 0.5.4 based on the comments in [this thread](https://github.com/grpc/grpc-node/issues/1027) and it did not help

Here is the new yarn.lock:
```
"@grpc/grpc-js@^0.5.2":
  version "0.5.4"
  resolved "https://registry.yarnpkg.com/@grpc/grpc-js/-/grpc-js-0.5.4.tgz#20ee7880e69805cbc63a1c54a3af02cc6475c850"
  integrity sha512-aY4fTCz7jq7oKFmfAeZVqGzMCR5I9NLdY9E2fJ70QtGXwlJnTaN6cnbRmCk23/aKPx9UHqOtk2lyjpN6LbAaxw==
  dependencies:
    semver "^6.2.0"

"@grpc/proto-loader@^0.5.0", "@grpc/proto-loader@^0.5.1":
  version "0.5.2"
  resolved "https://registry.yarnpkg.com/@grpc/proto-loader/-/proto-loader-0.5.2.tgz#c84f83be962f518bc303ca2d5e6ef2239439786c"
  integrity sha512-eBKD/FPxQoY1x6QONW2nBd54QUEyzcFP9FenujmoeDPy1rutVSHki1s/wR68F6O1QfCNDx+ayBH1O2CVNMzyyw==
  dependencies:
    lodash.camelcase "^4.3.0"
    protobufjs "^6.8.6"
```

**Comment by jakeleventhal** [comment]
If it is of any help, the way I use firestore is as follows:

I have a file called `firestore.js` that looks like this:
```javascript
import Firestore from '@google-cloud/firestore';

export default new Firestore();
```

I then import firestore and use it in other places like this:
```javascript
import firestore from './modules/firestore';

const updateThing = async (req, res) => {
  try {
    const { docId, updatePayload } = req.body;

    await firestore.doc(`Things/${docId}`).update(updatePayload);

    return res.status(200).send();
  } catch (err) {
    console.error(err);
    return res.status(500).send();
  }
};
```

I can probably have a temporary workaround by just creating a new Firestore instance from within each function body, but fixing the underlying problem is a more sustainable approach.

Also, my application doesn't have any event listeners. The only operations it uses are `get()`, `set()`, and `update()`

**Comment by schmidt-sebastian** [comment]
@jakeleventhal Thanks for the very detailed issue! Can you describe your work load a bit? It would help to know how many requests you are sending and if the memory usage tapers off once those requests are processed.

**Comment by murgatroid99** [comment]
I have a couple of observations here:

The string that is being constructed almost a million times looks like a generated variant of the protobuf.js `Message` constructor that is specific to various message types from the `.proto` files in the Firestore library. The most likely reason for that happening is that the `.proto` files are getting loaded repeatedly. Other memory usage from protobuf.js probably has the same cause. The library `@grpc/proto-loader` is essentially just a proxy for protobuf.js, so memory leaks there probably also have the same cause.

The `Http2Subchannel` objects from grpc are probably the result of repeatedly constructing new clients, and that version of the grpc library tends to keep channels and subchannels alive even when they are not being used. This is improved in `@grpc/grpc-js` version 0.6.x, mainly because subchannels with identical parameters will be reused.

**Comment by jakeleventhal** [comment]
@schmidt-sebastian So I have an API server that accepts HTTP requests from a another scheduling server. The scheduling server repeatedly sends jobs to the API to save user data.

There are many operations that occur but, for the most part, they can all be summarized as this:

1. Fetch a large data report from a 3rd party API
2. Parse that data to do one of or both of the following:
- a) Create a document with `set()`
- b) Modify an existing document with `update()`

There can be anywhere from 1-1000 or so firestore calls per minute per user. Each document that is updated is unique so there is no contention going on here.

I also have used log statements to see that functions are entering and exiting (and they all more or less follow a similar structure to what I wrote above). The scope of the firestore calls should be limited to just the function. Moreover, the firestore calls aren't throwing errors or anything and are updating documents successfully. Again, there are no event listeners anywhere in my code.

I have also tried not using a single instance of Firestore that I import throughout my code - I tried creating a new instance of Firestore within the body of each function that uses firestore. I tracked the number of instances of Firestore objects through memory snapshots and was able to see that the Firestore objects were successfully being deleted/deallocated at the end of each function; however, the metadata with all those strings/arrays/objects still persisted outside the scope of the functions.

I hope that clears up everything.

**Comment by swftvsn** [comment]
@jakeleventhal we've hit with this kinds of problems too in cloud functions: https://github.com/firebase/firebase-functions/issues/536 

The problems originate from grpc-js, which is still experimental.

What worked for us in the end was to downgrade our dependencies so it still uses the binary grpc which is much more mature and stable.

There's also a mechanism to force firestore to use binary grpc with the newer client version described here: https://github.com/firebase/firebase-functions/issues/536#issuecomment-531923813, however I have not tested that one as it is unsupported / untested if I understood correctly. That note also explains a bit why we are in this mess.

So, after grpc-js matures a bit, we’ll hop back to the latest versions.

**Comment by jakeleventhal** [comment]
@swftvsn what functionality will be lost/what differences will there be?

**Comment by jakeleventhal** [comment]
@swftvsn @schmidt-sebastian why is firestore not using the binary method of grpc until the other version is stable? what’s to be gained?

**Comment by swftvsn** [comment]
@jakeleventhal The reasoning to use grpc-js is explained in this comment: firebase/firebase-functions#536 

We didn't lose any functionality _we need_ when switching to earlier version, but of course we just checked everything we need. (We use typescript, so basically everything compiles -> you're good to go. ymmw etc. ofcourse.)

**Comment by schmidt-sebastian** [comment]
Providing a different GRPC implementation via the constructor argument (as described in https://github.com/firebase/firebase-functions/issues/536#issuecomment-531923813) is fully supported. All currently available features in Firestore are compatible with both `grpc` and `@grpc/grpc-js`.

@murgatroid99 Do you have a rough idea of where in our stack this problem might best be addressed?

**Comment by murgatroid99** [comment]
I'm only really familiar with my end of the stack, so I can only say that it's somewhere above that. Probably this firestore library. Loading the same `.proto` file repeatedly is not an intended usage of `@grpc/proto-loader`, so I recommend that you find a way to load it only once per process.

Similarly, constructing many clients for the same service and backend is not the intended usage of either grpc library, but that should be a much smaller problem with the new release of grpc-js that gax should be picking up soon.

**Comment by schmidt-sebastian** [comment]
I spent some more time on this, but I am still not able to pinpoint the leak. What is interesting though is that this only seems to happen with set()/update()/delete() and not with a get() call.

After 2000 operations, my memory Profile looks like this:

<img width="971" alt="Screen Shot 2019-10-01 at 2 50 05 PM" src="https://user-images.githubusercontent.com/19397744/66006337-32ff5400-e463-11e9-9c47-8cdec6f312fd.png">

The memory usage does not go down even when all operations resolve.

**Comment by jakeleventhal** [comment]
Is this confirmed to be fixed using the regular grpc library? i.e.
```javascript
import grpc from 'grpc'

const firestore = new Firestore({ grpc })
```

**Comment by merlinnot** [comment]
@schmidt-sebastian I don't know if it's related, but there is a similar issue in BigQuery: https://github.com/googleapis/nodejs-bigquery/issues/547

**Comment by schmidt-sebastian** [comment]
@jakeleventhal I see similar memory profiles with both `grpc` and `@grpc/grpc-js`.

**Comment by schmidt-sebastian** [comment]
It looks like the memory leak happens in one of our depend layers, so it might take a bit of work to track this down. The memory leak also exist when the Veneer client (which is an auto-generated SDK that we use for our request logic) is used directly. What is somewhat interesting is that we only seem to be leaking memory once we get responses from the backend - I see a steady increase of memory consumption only after requests are starting to get fulfilled.

With 2000 writes, FirestoreClient_innerApiCalls has a retained size of 99680 bytes that is mostly unaccounted for. There are also 2000 PromiseWrap and Promise objects on the stack that take up roughly this amount of memory.

**Comment by swftvsn** [comment]
Great to see the investigation resulting some fruit @schmidt-sebastian - and sorry that I jumped to conclusion and thus wrote bs above..

**Comment by jakeleventhal** [comment]
@schmidt-sebastian There is definitely a huge improvement using the legacy grpc (`new Firestore({ grpc })`).
These are tests run (with the exact same traffic/load

`grpc-js`
<img width="381" alt="Screen Shot 2019-10-02 at 8 57 08 PM" src="https://user-images.githubusercontent.com/15133828/66098408-5b687a80-e557-11e9-9d33-05312e2c33f8.png">


`grpc`
<img width="377" alt="Screen Shot 2019-10-02 at 8 56 40 PM" src="https://user-images.githubusercontent.com/15133828/66098413-61f6f200-e557-11e9-8ef4-7bc186298502.png">


You'll notice the memory is still climbing in the second example - but its very very minimal compared to the `grpc-js` version

**Comment by murgatroid99** [comment]
How are you measuring the memory usage? If you are using a JavaScript memory profiler, it may be omitting memory allocations by native code, which account for the majority of the memory usage of grpc.

**Comment by jakeleventhal** [comment]
The screenshots in the original post that show the memory profiler are using the chrome memory tool when running with the —inspect flag.
The screenshots provided of the memory graph are the memory usage of my application on GKE

**Comment by jakeleventhal** [comment]
Is this resolved on `2.4.0`?

**Comment by MysteryG** [comment]
@jakeleventhal 2.4.0 seemed to have resolved the issue on my side.

**Comment by jakeleventhal** [comment]
@MysteryG were you experiencing the same thing? and are/were you using the legacy grpc as a workaround

**Comment by MysteryG** [comment]
> @MysteryG were you experiencing the same thing? and are/were you using the legacy grpc as a workaround

@jakeleventhal I ended up reverting to version `2.2.9` as a temporary measure to avoid implementing a temporary work around.

**Comment by MysteryG** [comment]
@jakeleventhal Ignore last, we are still experiencing memory issues, and now loss in connection to Firestore.

**Comment by merlinnot** [comment]
I'm certain there's a memory leak in `@grpc/grpc-js` v0.6.x, which does not exist in `v0.5.4`. I have a reproduction for it, but it's not yet minimal (very small tho). I'd love to give access to this repo to some Google engineers, please let me know if you'd be interested. I can't share it publicly.

**Comment by schmidt-sebastian** [comment]
@MysteryG If you upgrade to @grpc/grpc-js 0.6.5 you should once again be able to connect to Firestore. See https://github.com/grpc/grpc-node/issues/1064

@merlinnot I would love to see your repro! I'll send you a message to the email listed on your profile.

**Comment by jakeleventhal** [comment]
Being that this is a paid service, why can’t @google-cloud/firestore be released with the legacy grpc until this is resolved? I’m sure many people are experiencing this issue.

**Comment by murgatroid99** [comment]
@merlinnot I am the maintainer of grpc-js. If you can share that repository with me that would be very helpful; I haven't seen any other information or reports of memory leaks in that version.

**Comment by merlinnot** [comment]
@schmidt-sebastian @murgatroid99 

I added you both as collaborators to the repository, hope GitHub sent you a link.

Reproduction steps:
```sh
git checkout repro
npm ci // Installs @grpc/grpc-js@0.6.
npm run test // Memory usage is logged, see it growing.
npm i @grpc/grpc-js@0.5
npm run test // Memory usage is logged, it's perfectly stable.
```

**Comment by murgatroid99** [comment]
@merlinnot I see the memory increasing in both cases, just by significantly smaller amounts with version 0.5. I believe you have mainly demonstrated that the `Client` class is more heavy-weight in version 0.6. In addition, some objects owned by the client may keep themselves alive for some amount of time, but we do not consider that to be a significant issue because users are expected to create and use a constant number of clients over the lifetime of a process, not create a new client for every request.

**Comment by jakeleventhal** [comment]
There is clearly a problem with 2.4.0
This is a graph of memory usage on my application over the last several days.
For the last week I had been using `@google-cloud/firestore` 2.3.0 with the legacy grpc in the constructor. I tried switching to 2.4.0 without the legacy grpc constructor and my memory usage skyrocketed.
<img width="551" alt="Screen Shot 2019-10-06 at 8 03 15 PM" src="https://user-images.githubusercontent.com/15133828/66282514-882cd280-e874-11e9-8b3f-ce6c9ffc9217.png">

**Comment by jakeleventhal** [comment]
I also was experiencing this error when fetching data on 2.4.0 with the default constructor
```
(node:43) UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 10.12.14.243:8080
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1126:14)
```

After deploying back to 2.3.0 with the default constructor my memory usage is way down again and i am no longer getting that error

**Comment by MysteryG** [comment]
Similarly I was running into connection errors after +- an hour of the service running on 2.4.0 
```js
Error: No connection established
    at Http2CallStream.call.on (/pipeline/source/node_modules/@grpc/grpc-js/build/src/call.js:68:41)
    at Http2CallStream.emit (events.js:203:15)
    at process.nextTick (/pipeline/source/node_modules/@grpc/grpc-js/build/src/call-stream.js:75:22)
    at process._tickCallback (internal/process/next_tick.js:61:11)
```

**Comment by jakeleventhal** [comment]
<img width="531" alt="Screen Shot 2019-10-06 at 11 02 19 PM" src="https://user-images.githubusercontent.com/15133828/66288270-5ffd9d80-e88d-11e9-860d-fec51ec6a90c.png">

Here you can see the results of before/after moving back to 2.3.0

**Comment by murgatroid99** [comment]
I have conducted my own test for possible memory leaks in `@grpc/grpc-js` version 0.6.6. For the test, I have a server running in another process. Then in the client, I call a single method, and when that method completes, I call it again in a loop. After every hundred completed methods, I run garbage collection and take a heap snapshot. I got these results with the following variations:

 1. **Construct a single client and use it to make all of the calls**. In this case, I observe no evidence of any memory leaks. **This is how we recommend using gRPC**.

 2. **Construct one client per call**. In this case I see a consistent leak of about 4KB per call+client, all various objects owned by Channels.

 3. **Construct one client per call, and close the client after the call completes**. In this case I see a leak of approximately a few hundred bytes per call. So far I can't tell what is using that memory.

 4. **Load the proto file and construct a client per call**. In this case I see a leak of about 50KB per call+client, similar to number 2, plus generated protobuf code. The value here likely depends strongly on the contents of the proto file(s).

 5. **Load the proto file and construct a client per call, and close the client after the call completes**. In this case I see a leak of about 45KB per call+client. I think this is just the same as number 4, with the same savings as in number 3.

If I run the same tests with `@grpc/grpc-js` version 0.5.4 I see the following:

 1. This case still leaks no memory.

 2. In this case I see a leak of about 6KB per call+client.

 3. In this case I see no memory leak, a difference from the 0.6.x

 4. In this case I see a leak of about 53KB per call+client. This is consistent with the same test case for 0.6.x, plus the slightly higher leak in number 2.

 5. In this case I see a leak of about 48KB per call+client. This is again consistent with the previous observations.

I am not really sure what is causing Firestore to use more memory when using `grpc-js` version 0.6, but I think it can be fixed by constructing a single client and using it to make all of the requests.

**Comment by schmidt-sebastian** [comment]
Thanks for the update!

Firestore is using multiple clients only if there are more than 100 concurrent requests. Since we have long-standing streams, we cannot just "wait" until the request count drops to zero before we can start another stream. 

We have logic that stops referencing clients that are no longer needed, but we don't do anything special to clean up their state (https://github.com/googleapis/nodejs-firestore/blob/a4efa097ef8de9ca4944356ab8767ddaa94c4188/dev/src/pool.ts#L174). From your investigating, it looks like we should at least be closing the GRPC channel manually. I will work on that for the time being.

**Comment by jakeleventhal** [comment]
@murgatroid99 i use one Firestore client

**Comment by schmidt-sebastian** [comment]
@jakeleventhal The Firestore client internally spins up multiple GRPC clients if needed.

**Comment by merlinnot** [comment]
Hi guys 👋 I believe I discovered a memory and network socket leak that might possibly be a root cause of this issue, or at least one of the causes, in a `@grpc/grpc-js` dependency.

I opened an issue with a reproduction, but I'd appreciate if you could validate that my reproduction yields the same results on your systems and that the reproduction steps are clear and easy to follow, so that gRPC team will be able to fix it more easily. Thanks in advance!

https://github.com/grpc/grpc-node/issues/1085

**Comment by schmidt-sebastian** [comment]
@merlinnot Thank you! In the meantime, I am also trying to see what it takes to make sure that our clients release all resources when they get garbage collected. This touches a lot of dependencies, so please bear with us.

**Comment by merlinnot** [comment]
Is there anything I can do to help? Today another update of @grpc/grpc-js was released, for which I can't easily find any significant memory leaks anymore. I could start making PRs against other repositories, but I'd need some guidelines on what do you mean by "a lot of dependencies" @schmidt-sebastian.

**Comment by jakeleventhal** [comment]
Any update on when this will be fixed? Im pinned at 2.3.0 and can't update to get the new features like the IN operator until this is fixed since it will cost too much $ to upgrade

**Comment by voxelbusters** [comment]
@jakeleventhal  May I know which tool/plugin you are using to profile the way you did?

**Comment by jakeleventhal** [comment]
@voxelbusters I was using the node --inspect flag with the chrome memory profiling tool

**Comment by bcoe** [comment]
@merlinnot that is good to know that you're no longer seeing memory leaks with the latest `@grpc/grpc-js`. We are gradually rolling this dependency out to some of the libraries I work on, I will keep you in the loop this week.

I know @alexander-fenster had released a `next` release of `gax` that pulls in some of the fixes.

**Comment by schmidt-sebastian** [comment]
@merlinnot "A lot of dependencies" refers to the updates we have to plumb through for https://github.com/googleapis/gapic-generator-typescript/issues/68

**Comment by merlinnot** [comment]
Thanks, I'm already following the other thread. I'm eager to help if you need anything (PRs, tests, ...) 👍

**Comment by urirahimi** [comment]
I'm experiencing this as well. Please fix

**Comment by jakeleventhal** [comment]
Is there an update on this? I'm literally paying for 5x more resources than I need right now because of this issue. It's starting to cost me a lot of money.

**Comment by schmidt-sebastian** [comment]
We have fixed numerous small leaks in `@grpc/grpc-js` (please upgrade to the latest version of google-gax or re-install Firestore). Next week, we are also planning to migrate the internal Firestore client to Typescript, which will allow us to close the GRPC channels on cleanup. We will release that the moment it is ready.

**Comment by urirahimi** [comment]
@schmidt-sebastian Any update on "next week" or did you have one of these weeks: https://www.mentalfloss.com/article/51370/why-our-calendars-skipped-11-days-1752

**Comment by urirahimi** [comment]
But in all seriousness, I'm spending a doggone fortune because of this leak..

**Comment by schmidt-sebastian** [comment]
I am very sorry, we are still working on this. We are making progress every day and I think I can see the finish line.

**Comment by urirahimi** [comment]
Is there any update on this? Bleeding cash for me.

**Comment by jakeleventhal** [comment]
Yeah, this has been nearly 3 months... come on...

**Comment by schmidt-sebastian** [comment]
We are basically done: https://github.com/googleapis/nodejs-firestore/pull/779 is waiting for a final review.
We also have some timeout settings that we need to fix in https://github.com/googleapis/nodejs-firestore/pull/805, but then we will release new client.

**Comment by schmidt-sebastian** [comment]
Version 3.0.0 should be released any minute now. With this version, we close the GRPC client when a channel becomes idle. If you request load frequently causes new channels to be spawn and released, this should reduce the memory burden described in this issue.

I am sorry and somewhat embarrassed by how long this took. I hope this solves the root issue described in here.

**Comment by bfitec** [comment]
Any idea how soon firebase-admin-node will be upgraded from 2.6v to start using 3.0v of nodejs-firestore? @schmidt-sebastian  I'm really glad a solution is in the pipe here, this has caused us a lot of grief, thank you for your efforts!

**Comment by jakeleventhal** [comment]
Experiencing over 15x reduction in memory footprint with 3.0.0 👍🏼

**Comment by schmidt-sebastian** [comment]
@bfitec Unfortunately, we may have another breaking change coming up shortly as Node 8 goes end of life end of the year. We still have to figure out what this means for us and if we have to bump the major version for Firestore again, but we don't want to bump the major version for Firebase Admin twice in short succession and hence we haven't updated Firebase Admin. 
Right now, we advise you to manually override the Firestore version for Firebase Admin (via `npm shrinkwrap` for example). Let us know if this is a huge hassle and we can reconsider.

**Comment by jmacedoit** [comment]
@schmidt-sebastian I changed `package.lock` to edit the version of `@google-cloud/firestore` to 3.0.0 on a project using firebase admin. However I continue to observe huge amounts of memory (+3Gb) 

In my `npm ls` output only  v3.0.0 is shown as used. I also get this message `npm ERR! invalid: @google-cloud/firestore@3.0.0 /Users/joaomacedo/Projects/firestore-cgd-gateway/node_modules/firebase-admin/node_modules/@google-cloud/firestore` but I guess that is to be inspected since I altered the dependency forcefully to 3.0.0.

**Comment by jmacedoit** [comment]
In our service we are listening for some dozens of thousands of streams. We noticed the more streams we need to be listening to the more memory our process needs. It is currently requiring +4Gb and increasing as new clients join. This should not happen as 98% of the streams have no messages being exchanged other that the first couple.  As I said above, we attempted forcing the 3.0.0 firestore cloud version by editing the package.lock.

**Comment by jmacedoit** [comment]
@schmidt-sebastian You advised shrinkwrap. Is using npm shrinkwrap for this case functionally different from changing package.lock?

**Comment by schmidt-sebastian** [comment]
@jmacedoit I added a branch to `firebase-admin-node` that should eliviate some of the installation pain. You can install as follows:

```
 "dependencies": {
    "firebase-admin": "git+https://github.com/firebase/firebase-admin-node.git#firestore-30"
  }
```

The bad news is that this might not address your specific memory concerns. GRPC only supports 100 concurrent operations at a time and as such, we need to create new GRPC clients for every 100 operations. The changes in this PR merely address the memory usage of GRPC clients that are no longer needed, but if you have a steady count of thousands of listeners, then some high memory usage is expected (at least for now).

**Comment by jmacedoit** [comment]
@schmidt-sebastian Any tips on how can this be addressed? The project in question is a chatboat gateway. It needs to listen to the streams of thousands of users (1 stream per user), in order to send the message to a core so that it is processed and the answer is placed back in the client stream. Naturally, most of the time the streams will have no activity. Is there anything that can be done to mitigate this?

**Comment by schmidt-sebastian** [comment]
Two options come to mind:
- If latency allows, use Cloud Functions to trigger an operation whenever a new message is written.
- Use the same collection ID as a parent to all messages and use a single listener via [Collection Groups](https://firebase.googleblog.com/2019/06/understanding-collection-group-queries.html).

If you have engineering resources to spare, we could also add an option to issue multiple listeners over the same stream. This is supported by the backend and currently used by the Android, iOS and Web SDK. We deliberately decided to use one stream per listeners as this provides isolation for all operations in the Server SDKs, but we could optionally bundle listeners together in the Node Server SDK. Unfortunately, this is a significant effort and will likely not be staffed by us in the near term.

As for your last question - yes, a new operation is spun once `onSnapshot` is called.

**Comment by jmacedoit** [comment]
@schmidt-sebastian Collection groups would be great, unfortunately the client applications are already deployed (and the collection names to which they write have names which differ for each user) and we need a short term solution to the problem. I guess we will have to go with the cloud functions route, although to be honest I prefer to avoiding cloud functions whenever possible...

**Comment by jakeleventhal** [comment]
Looks like this issue generated yet another issue 🤦🏻‍♂️
https://github.com/googleapis/nodejs-firestore/issues/829

**Comment by siddhant-mohan** [comment]
The problem is still not solved. I am using firebase functions 3.6.1.
My package.json file is:
`{
  "name": "functions",
  "description": "Cloud Functions for Firebase",
  "scripts": {
    "lint": "eslint .",
    "serve": "firebase serve --only v1",
    "shell": "firebase functions:shell",
    "start": "npm run shell",
    "deploy": "firebase deploy --only v1",
    "logs": "firebase functions:log"
  },
  "dependencies": {
    "body-parser": "^1.19.0",
    "camelcase": "^5.3.1",
    "es6-promise-pool": "^2.5.0",
    "express": "^4.17.1",
    "firebase-admin": "~8.0.0",
    "firebase-functions": "^3.6.1",
    "firebase-functions-helper": "^0.7.5",
    "glob": "^7.1.6",
    "react-native-uuid": "^1.4.9"
  },
  "engines": {
    "node": "10"
  },
  "devDependencies": {
    "eslint": "^5.12.0",
    "eslint-plugin-promise": "^4.0.1"
  },
  "private": true
}
`

My function file is:
`
'use strict';

const functions = require('firebase-functions');
const admin = require('firebase-admin');
const firebaseHelper = require("firebase-functions-helper");
const db = admin.firestore();

const promisePool = require('es6-promise-pool');
const PromisePool = promisePool.PromisePool;
// Maximum concurrent account deletions.
const MAX_CONCURRENT = 2;
var FieldValue = require('firebase-admin').firestore.FieldValue;


/**
 * Run once a day at midnight, to cleanup the users
 * Manually run the task here https://console.cloud.google.com/cloudscheduler
 */
exports.autoAcceptFriendship = functions.runWith({memory: '2GB', timeoutSeconds: '360'}).pubsub.schedule('every 30 minutes').onRun(async context => {
  // Fetch all user details.
  const pendingFriendships = await pendingFriendship();
  // Use a pool so that we delete maximum `MAX_CONCURRENT` users in parallel.
  const promisePool = new PromisePool(() => approveFriendship(pendingFriendships), MAX_CONCURRENT);
  await promisePool.start();
  console.log('All bot friendship approved');
});

/**
 * Deletes one inactive user from the list.
 */
async function approveFriendship(pendingFriendships) {
  if (pendingFriendships.length > 0) {
    const friendshipToApprove = pendingFriendships.pop();

    console.log('friendship accepted started for ', JSON.stringify(friendshipToApprove));

    //accept Friendship
  try {
    const friendship = {
      user1: friendshipToApprove.user1,
      user2: friendshipToApprove.user2,
      jointIds: [friendshipToApprove.user1, friendshipToApprove.user2].sort(),
      created_at: FieldValue.serverTimestamp(),
    };
    const newDoc = await firebaseHelper.firestore
      .createNewDocument(db, 'friendships', friendship);

    const deletedContact = await firebaseHelper.firestore
      .deleteDocument(db, 'pending_friendships', friendshipToApprove.id);
    console.log('friendship accepted');
  }
  catch (error) {
    console.log('error occured', error);
    }
  } else {
    return null;
  }
}

/**
 * Returns the list of all inactive users.
 */
async function pendingFriendship() {

  let date = new Date();

  date.setDate(date.getDate() - 1);

  console.log('date is ', date)

  //for prod
  let userIds = ['ZZ2qctmj2kTzBzFu6s1eywO4Lzd2', 'ZH1VLx9PapNwP70AGhG917CnreD3', 'O9o91D05g5ek7DUYPzCfSnenqMC2',
    '89vXTJlfWCd1yd9Bm34Daiiwsdv1', '0S8JmDFG2haPFsyI71hmJcAOSsl1', 'wIOWwAUfaSTPonThkN4rU1zy4uq2', 'pqrVRB1zhhdOoBcYvwGSGTmajCH2',
    'hjf7nq6dTmP5eseoIiuh1XV4air1', 'vUtLDsL7JvUMP4mxJi1eqiB2HXc2'];

  // //for staging
  // let userIds = ['n8jjShDDNfaQNOoiuYOp0rwPOG82', '6Y4MCrQQLYNKhy80VlEjQR2HsH33','IcLqDsK15dMYCHYOCTO5901CM2G2',
  //   '0W1OjdtmoXN890PRxxygJ14ZboA3', 'fN7X29HafRYryP61DnGSG9Cp2ZA3', 'hAG3ezNb9XeQ7JslKFDcPjirHzI2',
  //   'AwhKncRjjyVLOpSw97mB8nlXUaD3', 'AwhKncRjjyVLOpSw97mB8nlXUaD3', '1HUPxAQDd8cnfb7yB61Sd1INO222', 'pljEAbuLmIUm2bmQF9woqwRhJ1l2'];

  let p_friendship_array = [];

  let pend_friendship_data = await firebaseHelper.firestore.queryData(db, 'pending_friendships', [['created_at', '>=', date]]);


  let data = [];

  Object.keys(pend_friendship_data).forEach((key, Index) => {
    let val = pend_friendship_data[key];
    val = {...val, id: key}
    data.push(val)
  })

  p_friendship_array = data.filter(u => {
    return userIds.indexOf(u.user2) !== -1;
  });

  console.log('no of pend friendship request is ', p_friendship_array.length)

  return p_friendship_array;
}
`

I am getting error as:
**memory limit exceeded. Function invocation was interrupted**

**Comment by schmidt-sebastian** [comment]
@siddhant-mohan If you turn on logging (`setLogFunction`) we can check how many GRPC clients you are using. Given your usage of PromisePool, this should just be one client (which can handle 100 concurrent requests). The debug logs would help us verify.

**Comment by siddhant-mohan** [comment]
@schmidt-sebastian hey can you help me how to turn on logging to check GRPC clients my function is using? Sorry I am quite new to the firebase functions

**Comment by schmidt-sebastian** [comment]
It should be:

```
admin.firestore.setLogFunction(console.log);
```

**Comment by siddhant-mohan** [comment]
This is what i am getting as a setLogFunction output

![image](https://user-images.githubusercontent.com/8568200/81688459-3934c480-9477-11ea-967b-9a5ae6c08b1c.png)
